### DISCUSSION TITLE
is there any way to declare a effect dependent on another?

### DISCUSSION THREAD

**Discussion by dark-ether** [haskell-effectful/effectful#120] [Q&A]
i am trying to write a megaparsec effect however MonadParsec requires a MonadPlus instance i don't want to write a MonadPlus instance as that would create problems with the NonDet Effect, but ideally to use the MonadParsec instance it would only be necessary to require the megaparsec effect is there any way to make the NonDet Effect implicit in the Megaparsec effect?

**Comment by arybczak** [comment]
> is there any way to make the NonDet Effect implicit in the Megaparsec effect?

Somewhat. You can create a `runParser :: Eff (Parser e s : NonDet : es) a -> Eff es a` interpreter that'll also put relevant `NonDet` on the stack, but nothing prevents callers from overwriting your `NonDet` interpreter.

To be fair, `ParserT` has the same issue if someone uses `ParserT`, then puts another transformer on top of it that reimplements `MonadPlus`, but leaves `MonadParser` the same :shrug:
